File: leetcodeGo/sliding_window/[blind 150]_PermutationInString.go
import "fmt"

func areSame(a, b map[byte]int) bool {
    if len(a) != len(b) { return false }
    for k,v := range a {
        if val, ok := b[k]; !ok || val != v {
            return false
        }
    }
    return true
}

func checkInclusion(s1 string, s2 string) bool {
    ls1, ls2 := len(s1), len(s2)
    if ls1 > ls2 { return false } //edge case

    m1 := make(map[byte]int,3)
    m2 := make(map[byte]int,3)
    l,r := 0, len(s1)
    
    //populating m1 from s1 and m2 from first window of s2
    for i := 0; i < ls1; i++ {
        m1[s1[i]]++
        m2[s2[i]]++
    }
    //check if the first window is a permutation of s1
    if areSame(m1, m2) { return true }
    
    for r < ls2 {
        // Remove the character s2[l] that is sliding out of the window and increment l
        m2[s2[l]] -= 1
        if m2[s2[l]] == 0 { 
            delete(m2, s2[l]) 
        }
        l++
        //add new character s2[r] into the window and increment
        m2[s2[r]]++
        //check if the new window is a permutation of s1, and increment r
        if areSame(m1, m2) {
            return true
        }
        r++
    }
    return false
}

//TC: O(n) where n is length of s2
//SC: O(2*m) where m is length of s1 - for the hashmaps

File: leetcodeJS/sliding_window/[blind 150]_PermutationInString.rs
function areSame(m1: Object, m2: Object): boolean {
    for (const k in m1) {
        if (!(k in m2) || m2[k] !== m1[k]) {
            return false;
        }
    }    
    return true;
}

function checkInclusion(s1: string, s2: string): boolean {
    let [l, r] = [0, s1.length];
    let s2a = s2.split(''); 
    let s1a = s1.split('');
    let [m1, m2] = [{}, {}];
    
    for (let i=0; i<s1.length; i++) {
        if (s1[i] in m1) {
            m1[s1[i]]  += 1;
        } else {
            m1[s1[i]] = 1;
        }
        
        if (s2[i] in m2) {
            m2[s2[i]]  += 1;
        } else {
            m2[s2[i]] = 1;
        }
    }

    if (areSame(m1, m2)) {
        return true
    }

    while (r < s2.length) {
        m2[s2[l]]--;
        if (m2[s2[l]] === 0) {
            delete m2[s2[l]];
        }
        l++;

        if (s2[r] in m2) {
            m2[s2[r]]++;
        } else {
            m2[s2[r]] = 1;
        }
        r++;

        if (areSame(m1, m2)) {
            return true;
        }
    }
    return false;
